design/anonymousTokens.md

```diff
@@ -0,0 +1,104 @@
+# Auth2 Anonymous tokens
+
+## Problem statement
+
+Currently KBase users must log into KBase in order to view a narrative or download (DL) data from
+a narrative. We wish for users that are not logged in to be able to do both.
+
+## Background
+
+* When a Narrative docker container is spun up for a user, the narrative launcher expects a KBase
+  auth token and assigns the container to the user.
+* JSON workspace object DLs go through the Data Import Export (DIE) server, which appears
+  to handle anonymous requests correctly based on examination of the code (would need to be
+  tested).
+* Other DLs go through the Narrative Job Server (NJS), which requires a token - and all KBase
+  Software Development Kit (SDK) apps require a token - to create the DL package in Shock.
+  * DIE is responsible for DLing the package from Shock for the user in the case of
+    non-staging area DLs
+      * DIE appears to handle anonymous DLs correctly based on examination of the
+        code (would need testing).
+      * HOWEVER, the Shock node would need to be publicly readable, and no DL apps
+        make that change (because they all expect a token).
+
+## Document purpose
+
+Propose a preliminary design for KBase auth service support of anonymous tokens (ATs)
+that allow anonymous users to have a transient identity when interacting with KBase. Discuss how
+those tokens would interact with the rest of the KBase infrastructure and what changes would be
+necessary.
+
+## Anonymous tokens
+
+* The KBase auth service would be enhanced to offer AT support.
+* ATs would have the same fields as regular tokens except that the the `type` would
+  always be `Anonymous`.
+* ATs would have a random user name assigned upon creation.
+  * Could be a concatenated dictionary style name for mnemonic purposes
+    (e.g. [https://gfycat.com/shoddydismalamoeba](https://gfycat.com/shoddydismalamoeba)).
+      * Need to check collision frequency.
+* ATs would exist in a separate token and user namespace from standard tokens.
+* ATs could not be used to create other tokens.
+* ATs would be prefixed with a string to indicate they are anonymous
+  (e.g. `***ANONYMOUS***-[remainder of token goes here]`)
+* ATs would be created, introspected, and revoked with separate APIs from normal tokens.
+  * Token creation requires no authentication.
+  * Test mode would support creation and introspection (token revocation is not supported in
+    general via test mode and hasn't been needed so far)
+* The administration UI would have AT and user lookup functionality and the ability
```

Comment:
It might be a good idea to have a blacklist of IPs, IP ranges - if a client matches the blacklist which we won't issue an anonymous token. The full IP blacklists that threat intelligence sources provide tend to be huge, I don't think we need to support those, but if we see a lot of activity or are warned about such activity, we should have a way to block IPs or entire ranges of IPs.
http://iplists.firehol.org/?ipset=et_block

Reply:
Would it make sense for this to be implemented in the auth service or at the nginx level?

Reply:
Ideally it would be at the nginx level, the problem is that this incurs the overhead of a blacklist IP check on every http connection that comes into the system. I have an nginx feature branch that supports this (and more) but I haven't have a chance to test it for the performance impact.
https://github.com/kbase/nginx/tree/modsecurity

Reply:
Added a note regarding this